test/src/test/java/org/corfudb/infrastructure/logreplication/LogEntryWriterTest.java

```diff
@@ -0,0 +1,403 @@
+        numOpaqueEntries = 3;
+        topologyConfigId = 5;
+        utils = new TestUtils();
+    }
+
+    /**
+     * This test verifies that Log Entry(delta) updates are applied on the Sink with transaction boundary preserved.
+     * Each OpaqueEntry maps to a transaction on the Source.  Hence, each should be applied on the Sink as a separate
+     * transaction.
+     */
+    @Test
+    public void testLogEntryApplyWithExpectedTx() {
+        LogReplicationEntryMsg lrEntryMsg = utils.generateLogEntryMsg(1, numOpaqueEntries, Address.NON_ADDRESS,
+            topologyConfigId, Address.NON_ADDRESS);
+
+        Map<LogReplicationMetadataType, Long> metadataMap = constructMetadataMgrMap(topologyConfigId,
+            Address.NON_ADDRESS, Address.NON_ADDRESS, Address.NON_ADDRESS);
+        updateMetadataManagerMap(metadataMap, false);
+
+        // Verify that the message containing the list of opaque entries was applied successfully
+        Assert.assertTrue(logEntryWriter.apply(lrEntryMsg));
+
+        verifyNumberOfTx(numOpaqueEntries);
+
+        // Construct the expected arguments and order with which the metadata updates will be written.  There will be 4
+        // updates - 3 for LAST_LOG_ENTRY_APPLIED and 1 for LAST_LOG_ENTRY_BATCH_PROCESSED
+        List<TxnContext> txnContexts = new ArrayList<>();
+        List<LogReplicationMetadataType> metadataTypes = new ArrayList<>();
+        List<Long> timestamps = new ArrayList<>();
+
+        for(int i = 0; i < numOpaqueEntries + 1; i++) {
+            txnContexts.add(txnContext);
+        }
+
+        for(int i = 0; i < numOpaqueEntries + 1; i++) {
+            // The first 3 metadata updates will be for LAST_LOG_ENTRY_APPLIED
+            if (i < numOpaqueEntries) {
+                metadataTypes.add(LogReplicationMetadataType.LAST_LOG_ENTRY_APPLIED);
+                timestamps.add((long)i+1);
+            } else {
+                // The last one will be for LAST_LOG_ENTRY_BATCH_PROCESSED
+                metadataTypes.add(LogReplicationMetadataType.LAST_LOG_ENTRY_BATCH_PROCESSED);
+                timestamps.add((long)i);
+            }
+        }
+        verifyMetadataAppliedAndOrder(numOpaqueEntries+1, txnContexts, metadataTypes, timestamps);
+    }
+
+    /**
+     * This test simulates the case that the same payload is received by LogEntryWriter twice and verifies that it is
+     * applied only once.
+     */
+    @Test
+    public void testLogEntryRedundantApply() {
+        testLogEntryApplyWithExpectedTx();
+
+        // Send the same payload again
+        LogReplicationEntryMsg lrEntryMsg = utils.generateLogEntryMsg(1, numOpaqueEntries, Address.NON_ADDRESS,
+            topologyConfigId, Address.NON_ADDRESS);
+
+        Map<LogReplicationMetadataType, Long> metadataMap = constructMetadataMgrMap(topologyConfigId, numOpaqueEntries,
+            numOpaqueEntries, Address.NON_ADDRESS);
+        updateMetadataManagerMap(metadataMap, true);
```

Comment:
reset() is resetting the metadataManager, hence it resets the txnContext returned from it.

The resetting of metadataMap with the timestamps is required because
1) we want to verify against different metadata
2) if txnContext is reset(due to metadataManager reset), all metadata fetched using it will be null.